pkg/prowloader/prow.go

```diff
@@ -0,0 +1,292 @@
+	"gorm.io/gorm/clause"
+	"k8s.io/klog"
+
+	"github.com/openshift/sippy/pkg/apis/junit"
+	"github.com/openshift/sippy/pkg/apis/prow"
+	sippyprocessingv1 "github.com/openshift/sippy/pkg/apis/sippyprocessing/v1"
+	"github.com/openshift/sippy/pkg/db"
+	"github.com/openshift/sippy/pkg/db/models"
+	"github.com/openshift/sippy/pkg/prowloader/gcs"
+	"github.com/openshift/sippy/pkg/testidentification"
+)
+
+const TestStatusFlake = 13
+const TestStatusFail = 12
+const TestStatusSuccess = 1
+
+const prowURL = "https://prow.ci.openshift.org/prowjobs.js?var=allBuilds&omit=annotations,labels,decoration_config,pod_spec"
+
+type ProwLoader struct {
+	dbc                 *db.DB
+	bkt                 *storage.BucketHandle
+	bktName             string
+	prowJobCache        map[string]*models.ProwJob
+	prowJobRunCache     map[uint]bool
+	prowJobRunTestCache map[string]uint
+	variantManager      testidentification.VariantManager
+	suiteCache          map[string]uint
+}
+
+func New(dbc *db.DB, gcsClient *storage.Client, gcsBucket string, variantManager testidentification.VariantManager) *ProwLoader {
+	bkt := gcsClient.Bucket(gcsBucket)
+
+	return &ProwLoader{
+		dbc:                 dbc,
+		bkt:                 bkt,
+		bktName:             gcsBucket,
+		prowJobRunCache:     loadProwJobRunCache(dbc),
+		prowJobCache:        loadProwJobCache(dbc),
+		prowJobRunTestCache: make(map[string]uint),
+		suiteCache:          make(map[string]uint),
+		variantManager:      variantManager,
+	}
+}
+
+func loadProwJobCache(dbc *db.DB) map[string]*models.ProwJob {
+	prowJobCache := map[string]*models.ProwJob{}
+	var allJobs []*models.ProwJob
+	dbc.DB.Model(&models.ProwJob{}).Find(&allJobs)
+	for _, j := range allJobs {
+		if _, ok := prowJobCache[j.Name]; !ok {
+			prowJobCache[j.Name] = j
+		}
+	}
+	klog.V(4).Infof("job cache created with %d entries from database", len(prowJobCache))
+	return prowJobCache
+}
+
+// Cache the IDs of all known ProwJobRuns for this job. Will be used to skip job run and test
+// results we've already processed.
+func loadProwJobRunCache(dbc *db.DB) map[uint]bool {
+	prowJobRunCache := map[uint]bool{} // value is unused, just hashing
+	knownJobRuns := []models.ProwJobRun{}
+	ids := make([]uint, 0)
+	dbc.DB.Select("id").Find(&knownJobRuns).Pluck("id", &ids)
+	for _, kjr := range ids {
+		prowJobRunCache[kjr] = true
+	}
+
+	return prowJobRunCache
+}
+
+func (pl *ProwLoader) LoadProwJobsToDB() error {
+	allowedJobRegex := []*regexp.Regexp{
+		regexp.MustCompile(`pull-ci-openshift-.*-(master|main)-e2e-.*`), // For now let's just get master/main presubmits in the openshift org
+	}
+
+	jobJSON, err := fetchJobJSON(prowURL)
+	if err != nil {
+		return err
+	}
+	prowJobs, err := jobJSONToProwJob(jobJSON)
+	if err != nil {
+		return err
+	}
+
+	for _, pj := range prowJobs {
+		for _, re := range allowedJobRegex {
+			if re.MatchString(pj.Spec.Job) {
+				err := pl.prowJobToJobRun(pj)
+				if err != nil {
+					return err
+				}
+				break
+			}
+		}
+	}
```

Comment:
Could you give a test with this running and watching the res memory use?

Reply:
There are some spikes of up to 1G:

![rss](https://user-images.githubusercontent.com/429763/169076994-5effb33e-f19f-4a52-ab30-0b7601a68667.svg)